lib/system/web/src/keyboard_engine.rs

```diff
@@ -0,0 +1,122 @@
+use std::collections::{HashMap, BTreeSet};
+
+use web_sys::{KeyboardEvent};
+use wasm_bindgen::JsCast;
+use wasm_bindgen::prelude::*;
+
+use crate::*;
+
+type Bindings = HashMap<BTreeSet<u32>, Vec<Box<dyn FnMut()>>>;
+
+/// Keyboard Engine that handles user keyboard combinations
+///
+/// Keyboard Engine stores bindings as HashMap:
+/// {
+///     [keyA, keyB]:          [Callback1, Callback2],
+///     [keyA, keyB, keyCtrl]: [Callback1],
+/// }
+/// when the combination is pressed related callbacks are invoked
+pub struct KeyboardEngine {
+    pub bindings: Rc<RefCell<Bindings>>,
+    pub state:    Rc<RefCell<BTreeSet<u32>>>,
+    pub target:   web_sys::EventTarget
+}
+
+impl KeyboardEngine {
+    pub fn new(target: &web_sys::EventTarget) -> Result<Self> {
+        let state = Rc::new(RefCell::new(BTreeSet::new()));
+        let bindings = Rc::new(RefCell::new(HashMap::new()));
+
+        KeyboardEngine::init(target, Rc::clone(&bindings), Rc::clone(&state));
+
+        Ok(Self {
+            bindings,
+            state,
+            target: target.clone()
+        })
+    }
+
+    pub fn from_tag_name(html_tag: &str) -> Result<Self> {
+        let element = document()?
+            .get_elements_by_tag_name(html_tag)
+            .item(0).ok_or_else(|| Error::missing("there is no such element"));
+
+        KeyboardEngine::new(&element.unwrap())
+    }
+
+    fn init (
+        target:   &web_sys::EventTarget,
+        bindings: Rc<RefCell<Bindings>>,
+        state:    Rc<RefCell<BTreeSet<u32>>>
+    ) {
+        {
+            let state = Rc::clone(&state);
```

Comment:
There is a lot of strange code here. Here for example you take `state` ownership in the function and the first thing you do is you clone the RC. It doesnt make sense because youve got the ownership, right?

Reply:
I did it here cause I then move the state into the first callback, but I need it in the second one